### PR TITLE
Bugfix/real incremental diff based rendering

### LINKED ISSUE #408: Rerender loses focus of input
## Are you reporting a bug or a feature?

 - [x] Bug
 - [ ] Feature

## Expected Behavior

All components should re-render in an incremental manner, meaning:
- doing just updates und **NOT** replacing whole DOM sub trees -> DOM diffing
- be scoped at component level - should only mess with it's **local-DOM**
- keep intrinsic state of interactive HTML elements like `<input />`

## Actual Behavior

The whole DOM sub-tree is removed and re-written, this leads all interactive HTML elements to starts at it's initial state and in case of `<input />` loses it's focus during user interacting with it.

**Note:** `shouldUpdateCallback()` possibly mitigates this problem, but it's not a complete solution.

## Steps to Reproduce the Problem

react-input-loses-focus-within-ce

### LINKED ISSUE #416: Namespaced attributes aren't updated incrementally
## Are you reporting a bug or a feature?

 - [x] Bug
 - [ ] Feature

## Expected Behavior

All attributes including namespaced attributes should update on incremental rendering.

## Steps to Reproduce the Problem

Namespaced attributes do not update.

1. Open Firefox
2. Go to https://axa-ch.github.io/patterns-library/atoms.html#a-icon

**Chrome (Using new SVG `<use href/>`)**
incremental-icon-chrome

**Firefox (Using old SVG `<use xlink:href/>`)**
incremental-icon-firefox-stops

## Specifications

 - Browser: Firefox
 - Browser Version: Latest
 - Nanomorph: https://github.com/choojs/nanomorph/issues/101
 - Morphdom: https://github.com/patrick-steele-idem/morphdom/issues/124

### PR DESCRIPTION
**Pull request by AndyOGo**
Fixes #408 
fixes #416 

Changes proposed in this pull request:

 - ~~installs [`nanomorph`](https://github.com/choojs/nanomorph)~~ We morph by ourselves
 - fixes DOM subtree replacements cause loos of intrinsic `<input />` states like `focus`

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update
 
 # How Has This Been Tested?
 
 

Needs tests in every browser.
- Latest Chrome works
   incremental-icon-chrome
- Latest firefox works partially, stops after first incremental update 🤔 
   https://github.com/choojs/nanomorph/issues/101
   incremental-icon-firefox-stops

 # Checklist:
 
 - [x] My code follows the style guidelines of this project
 - [x] I have performed a self-review of my own code
 - [x] I have commented my code, particularly in hard-to-understand areas
 - [x] I have made corresponding changes to the documentation
 - [x] My changes generate no new warnings
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [ ] New and existing unit tests pass locally with my changes
 - [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by AndyOGo** [comment]
@LucaMele 
@TheDadi 
the stroke is broken by enabling incremental rendering 😞 

morph-stroke-broken

**Comment by AndyOGo** [comment]
stroke fixed

File: src/js/abstract/base-component.js [hunk 1]
```diff
@@ -220,39 +221,49 @@ export default class BaseComponent extends HTMLElement {
    * @return {type}  description
    */
   render() { // eslint-disable-line
-    const { _hasRendered: initial } = this;
+    const { _hasRendered: notInitial } = this;
 
     if (ENV !== PROD) {
-      lifecycleLogger(this.logLifecycle)(`willRenderCallback -> ${this.nodeName}#${this._id} <- initial: ${!initial}`);
+      lifecycleLogger(this.logLifecycle)(`willRenderCallback -> ${this.nodeName}#${this._id} <- initial: ${!notInitial}`);
```
Comment (LucaMele) [inline src/js/abstract/base-component.js]: not notInitial is more readable?

File: src/js/abstract/base-component.js [hunk 2]
```diff
@@ -468,3 +499,18 @@ export default class BaseComponent extends HTMLElement {
...
+
+/**
+ * Make sure that another piece of code is/can managing that part of the DOM tree.
+ *
+ * @link https://github.com/choojs/nanomorph#caching-dom-elements
+ * @param node
+ */
+// @todo: ideally this code is only attached during morphing phase
+function isSameNodeOnce(node) {
+  node.isSameNode = isSameNodeStopMorph;
```
Comment (LucaMele) [inline src/js/abstract/base-component.js]: `node.isSameNode = () => true;`

File: src/js/abstract/component-morph.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
+import morph from './morph';
```
Comment (AndyOGo) [inline src/js/abstract/component-morph.js]: we will create it's own npm package for morphin in the future

**Comment by AndyOGo** [comment]
@LucaMele 
All fixed, Firefox works, SVG icon with xlink works, etc.

**Comment by AndyOGo** [comment]
Successfully tested in Edge 15 and 16 too

**Comment by LucaMele** [comment]
@AndyOGo ur a coder <GOD>

File: src/js/abstract/component-morph.js (hunk 3 above)
Comment (LucaMele) [reply src/js/abstract/component-morph.js]: Maybe add a todo in the file directly and also WHY we did copy/modify the nanomorph one

**Comment by AndyOGo** [comment]
@LucaMele 
Thank you very much courageous testing:)

hehe, thanks for your very friendly commendation :smiley:

**Pull request merged**